kernel/pci/src/lib.rs

```diff
@@ -221,74 +263,98 @@ impl PciLocation {
     /// The least two significant bits of offset are masked, so it's 4-byte aligned addressing,
     /// which makes sense since we read PCI data (from the configuration space) in 32-bit chunks.
     fn pci_address(self, offset: u8) -> u32 {
-        ((self.bus  as u32) << 16) | 
-        ((self.slot as u32) << 11) | 
-        ((self.func as u32) <<  8) | 
-        ((offset as u32) & (PCI_CONFIG_ADDRESS_OFFSET_MASK as u32)) | 
-        0x8000_0000
+        BASE_OFFSET |
+        ((self.bus  as u32) << 16) |
+        ((self.slot as u32) << 11) |
+        ((self.func as u32) <<  8) |
+        (offset as u32)
     }
 
-    /// read 32-bit data at the specified `offset` from the PCI device specified by the given `bus`, `slot`, `func` set.
-    fn pci_read_32(&self, offset: u8) -> u32 {
-        unsafe { 
-            PCI_CONFIG_ADDRESS_PORT.lock().write(self.pci_address(offset)); 
+    fn pci_read_32_addr(&self, address: u32) -> u32 {
+        #[cfg(target_arch = "x86_64")] {
+            let shift = (address & (!PCI_CONFIG_ADDRESS_OFFSET_MASK)) * 8;
+            let aligned_address = address & PCI_CONFIG_ADDRESS_OFFSET_MASK;
+            unsafe { 
+                PCI_CONFIG_ADDRESS_PORT.lock().write(aligned_address);
+            }
+            PCI_CONFIG_DATA_PORT.lock().read() >> shift
+        }
+
+        #[cfg(target_arch = "aarch64")] {
+            let config_space = PCI_CONFIG_SPACE.lock();
+            let config_space = config_space.deref().as_ref().unwrap();
+            let address = address as usize;
+            let slice = &config_space[address..][..size_of::<u32>()];
+            let mut value = [0; 4];
+            value.copy_from_slice(slice);
```

Comment:
hmm, i'm not sure that this is correct or even sound. Two things:
1. Don't you have to write to the ADDRESS port and then read from the DATA port, just like on x86 with port I/O?
2. You're accessing the cfg space like it's just a static chunk of memory, but it's not. At the very least, you need to use volatile operations to ensure the compiler doesn't remove/optimize out the memory reads here. The config space can be written to and read from interactively, i.e., its values may change depending on what you write to it. So you need to ensure that every read and write operation is done using the proper two-step sequence of instructions (write the address, then read/write the data).

Reply:
Also, I think this may be why you're getting invalid data out of the PCI cfg space in some cases ---- this shouldn't work, but it did happen to work sometimes. Fixing this may also fix the issue of `highmem=on` being required.

Reply:
> Also, I think this may be why you're getting invalid data out of the PCI cfg space in some cases ---- this shouldn't work, but it did happen to work sometimes. Fixing this may also fix the issue of `highmem=on` being required.

I doubt it, but I'll try.

> 1. Don't you have to write to the ADDRESS port and then read from the DATA port, just like on x86 with port I/O?

Nope

> 2. You're accessing the cfg space like it's just a static chunk of memory, but it's not. At the very least, you need to use volatile operations to ensure the compiler doesn't remove/optimize out the memory reads here.

You're fully right.

Reply:
Fun fact, I thought the high PCIE address was so high because of the highmem options but it's not. With and without the option, the address still is the same. So, even if it's unrelated to these reads and writes (still, thanks for the idea), we no longer need the highmem options x)

Reply:
oh interesting. Okay cool, lmk how it goes.